### DISCUSSION TITLE
Using AWS Controllers for Kubernetes (ACK) for AWS resource reconciliation

### DISCUSSION THREAD

**Discussion by mikestef9** [kubernetes-sigs/cluster-api-provider-aws#3306] [Proposals / ADRs]
**Background**
AWS Controllers for Kubernetes (ACK) enables users to define and use AWS service resources using Kubernetes native API resources.

Kubernetes Cluster API for AWS (CAPA) provides consistent deployment and day 2 operations of "self-managed" and EKS Kubernetes clusters on AWS.

ACK service controllers are a direct mapping to AWS service APIs, whereas CAPA operates at a higher level of abstraction than ACK, for example, creating a VPC if one is not specified as part of a cluster definition.

There is duplication of code and effort between ACK controllers and the lower level of CAPA that ensures AWS resources are kept in sync with desired state. 

**Proposal**
ACK now has service controllers for most AWS APIs that CAPA requires:

* **IAM**: https://github.com/aws-controllers-k8s/iam-controller
* **EC2/VPC**: https://github.com/aws-controllers-k8s/ec2-controller
* **KMS**: https://github.com/aws-controllers-k8s/kms-controller
* **Managed AWS Control Plane (EKS)**: https://github.com/aws-controllers-k8s/eks-controller

We propose a refactor of CAPA to import ACK controllers as packages, replacing existing CAPA ↔ AWS reconciliation code. ACK controllers are compatible as Go modules, and CAPA Go packages should be able to take a dependency on the controller packages, and then systematically declare ACK resources. We don’t believe there is anything specific about ACK code paths that requires they need to be run as ACK controllers.

Another potential approach is to generate (using the ack-generate tool) the CAPA resource manager and SDK linkage code (which is similar to how ACK does this for Crossplane's [provider-aws](https://github.com/crossplane/provider-aws)).

**More info on ACK:**

* [Documentation](https://aws-controllers-k8s.github.io/community/docs/community/overview/)
* [Launch blog](https://aws.amazon.com/blogs/containers/aws-controllers-for-kubernetes-ack/)

**Additional Notes**

* CAPA and ACK both use [AWS SDK for Go v1](https://github.com/aws/aws-sdk-go).
* ACK controllers CAPA may require that are not built yet: 
    * ELBv2 https://github.com/aws-controllers-k8s/community/issues/481
    * EC2 Auto Scaling groups https://github.com/aws-controllers-k8s/community/issues/588
* Today, ACK runs as separate controllers, and each require an IAM role specific to that service. In this proposal, ACK code would be assuming the IAM role associated with the CAPA controller, which would have the broader service permissions needed overall by CAPA.

**Comment by richardcase** [comment]
I like the idea of adopting ACK...especially the part about removing duplication of code/effort.  

A few questions/comments come to mind initially:

1. Is there any timeframe on building the ELBv2 / EC2 ASG controllers?
2. How would we integrate this with [our multi-tenancy approach](https://cluster-api-aws.sigs.k8s.io/topics/multitenancy.html)? I'm guessing that if we use the resource managers directly (i.e. as packages) then we can pass the AWS SDK session we create in CAPA to the manager? (as an example [here](https://github.com/aws-controllers-k8s/ec2-controller/blob/main/pkg/resource/vpc/manager_factory.go#L48))
3. Are there any benefits to using the other approach you mention (i.e. generate the CAPA resource manager)?
4. It adds another set of dependencies for us, so it would be good to explore the support model and future plans for ACK

I will think some more on this.

@sedefsavas - this could be very interesting, especially if we can use the controllers as packages.

**Comment by sedefsavas** [comment]
Excited to see an ADR for this.
This has been in our minds for a long time, it would be great to start a discussion to find out answers to Richard's points.

We can go over this during our office hours, unfortunately it overlaps with ACK office hours.

**Reply by richardcase** [reply]
I can go to the ACK office hours this Monday if you are happy to host the CAPA office hours? I can then join the CAPA meeting afterwards.

**Comment by sedefsavas** [comment]
Another think we need to consider is we only make the AWS API fields that we support available to the user. When say through ACK if we create an ELB, all the fields will be available to be changed by the user.

**Reply by richardcase** [reply]
It we use it as a package then we still control what we expose to the user via our API.

**Reply by sedefsavas** [reply]
Yes, but since AWS services are now individual CRDs, users can go and change Kubernetes CR and we don't have any control over which fields could be modified as guessing ACK has its own webhooks.

**Reply by richardcase** [reply]
Interesting, i had a different view in my head. 

@mikestef9 - it would be good to dive a bit deeper into the proposal. Would it be possible to arrange a chat to discuss?

**Comment by mikestef9** [comment]
We've put some thoughts together on ACK side about making controllers more easily importable into other projects https://github.com/aws-controllers-k8s/community/issues/1299

**Reply by richardcase** [reply]
Great, thanks @mikestef9 . I'll take a read

**Comment by RSE132** [comment]
@mikestef9 Any further update on the integration ?

**Reply by richardcase** [reply]
This is not going to be happening in the near future.
